sql/analyzer/rules.go

```diff
@@ -531,14 +534,82 @@ func resolveNaturalJoins(ctx *sql.Context, a *Analyzer, n sql.Node) (sql.Node, e
 
 		projections := append(append(common, left...), right...)
 
-		return plan.NewProject(
+		transformed = plan.NewProject(
 			projections,
 			plan.NewInnerJoin(
 				join.Left,
 				join.Right,
 				expression.JoinAnd(conditions...),
 			),
-		), nil
+		)
+
+		return transformed, nil
+	})
+
+	if err != nil || transformed == nil {
+		return node, err
+	}
+
+	var nodesToUnresolve = []sql.Node{}
+	plan.Inspect(node, func(iNode sql.Node) bool {
+		if iNode == transformed || iNode == nil {
+			return false
+		}
+
+		nodesToUnresolve = append(nodesToUnresolve, iNode)
+		return true
+	})
+
+	if len(nodesToUnresolve) == 0 {
+		return node, err
+	}
+
+	return node.TransformUp(func(node sql.Node) (sql.Node, error) {
+		var found bool
+		for _, nu := range nodesToUnresolve {
+			if nu.String() == node.String() {
```

Comment:
`reflect.DeepEqual(nu, node)` it's probably less expensive than this and more bullet-proof than relying on the String method of the nodes